src/modules/fancyzones/lib/ZoneSet.cpp

```diff
@@ -196,6 +286,256 @@ IFACEMETHODIMP_(void) ZoneSet::MoveSizeEnd(HWND window, HWND zoneWindow, POINT p
+{
+    if (type == JSONHelpers::ZoneSetLayoutType::PriorityGrid && zoneCount <= 11)
+    {
+        CalculateUniquePriorityGridLayout(workArea, zoneCount, spacing);
+        return;
+    }
+
+    int rows = 1, columns = 1;
+    while (zoneCount / rows >= rows)
+    {
+        rows++;
+    }
+    rows--;
+    columns = zoneCount / rows;
+    if (zoneCount % rows == 0)
+    {
+        // even grid
+    }
+    else
+    {
+        columns++;
+    }
+
+    JSONHelpers::GridLayoutInfo gridLayoutInfo(JSONHelpers::GridLayoutInfo::Minimal{ .rows = rows, .columns = columns });
+
+    for (int row = 0; row < rows; row++)
+    {
+        gridLayoutInfo.rowsPercents()[row] = C_MULTIPLIER / rows;
+    }
+    for (int col = 0; col < columns; col++)
+    {
+        gridLayoutInfo.columnsPercents()[col] = C_MULTIPLIER / columns;
+    }
+
+    for (int i = 0; i < rows; ++i)
+    {
+        gridLayoutInfo.cellChildMap()[i] = std::vector<int>(columns);
+    }
+
+    int index = 0;
+    for (int col = columns - 1; col >= 0; col--)
+    {
+        for (int row = rows - 1; row >= 0; row--)
+        {
+            gridLayoutInfo.cellChildMap()[row][col] = index++;
+            if (index == zoneCount)
+            {
+                index--;
+            }
+        }
+    }
+    CalculateGridZones(workArea, gridLayoutInfo, spacing);
+}
+
+void ZoneSet::CalculateUniquePriorityGridLayout(Rect workArea, int zoneCount, int spacing) noexcept
+{
+    CalculateGridZones(workArea, predefinedPriorityGridLayouts[zoneCount - 1], spacing);
+}
+
+void ZoneSet::CalculateCustomLayout(Rect workArea, const std::wstring& customZoneSetFilePath, int spacing) noexcept
+{
+    wil::unique_cotaskmem_string guuidStr;
+    if (SUCCEEDED_LOG(StringFromCLSID(m_config.Id, &guuidStr)))
+    {
+        JSONHelpers::FancyZonesDataInstance().GetCustomZoneSetFromTmpFile(customZoneSetFilePath, guuidStr.get());
+        const auto& zoneSet = JSONHelpers::FancyZonesDataInstance().GetCustomZoneSetsMap()[guuidStr.get()];
+
+        if (zoneSet.type == JSONHelpers::CustomLayoutType::Canvas)
+        {
+            const auto& zoneSetInfo = std::get<JSONHelpers::CanvasLayoutInfo>(zoneSet.info);
+            for (const auto& zone : zoneSetInfo.zones)
+            {
+                int x = zone.x;
+                int y = zone.y;
+                int width = zone.width;
+                int height = zone.height;
+                DPIAware::Convert(NULL, x, y);
+                DPIAware::Convert(NULL, width, height);
+
+                RECT focusZoneRect{ x, y, x + width, y + height };
+                AddZone(MakeZone(focusZoneRect));
+            }
+        }
+        else if (zoneSet.type == JSONHelpers::CustomLayoutType::Grid)
+        {
+            CalculateGridZones(workArea, std::get<JSONHelpers::GridLayoutInfo>(zoneSet.info), spacing);
+        }
+    }
+}
+
+void ZoneSet::CalculateGridZones(Rect workArea, JSONHelpers::GridLayoutInfo gridLayoutInfo, int spacing)
+{
+    int gutter = spacing;
```

Comment:
Same as above, no need for `gutter`, `spacing` value can be used to handle all zone edges (between zones, and between zone and edge of work area).